### PR TITLE
Pass norm argument to spy

### PR DESCRIPTION
**Pull request by bakesbasha**
## PR Summary
I worked on a fix to #18917 with @kristian-cho by adding an argument for 'norm' as suggested on the issue page.

Before fix:
```
 ret = self.imshow(mask, interpolation='nearest', aspect=aspect, 
                   origin=origin, **kwargs) 
```
After fix:
```
ret = self.imshow(mask, norm=mcolors.NoNorm(), interpolation='nearest', aspect=aspect,
                              origin=origin, **kwargs)
```

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A ] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by jklymak** [comment]
If possible, I'd add a test.  I'm not actually clear what NoNorm does.

**Comment by bakesbasha** [comment]
Sure, I'm kind of new to contributing though, not sure how I should write the test. Any tips?

**Comment by jklymak** [comment]
Have a look at our developer guide!

**Comment by tacaswell** [comment]
Specifically have a look at https://matplotlib.org/devdocs/devel/testing.html?highlight=check_figures_equal#writing-an-image-comparison-test .  I think a good test would be to compare the same data into spy and an `imshow` with the `vmin=0, vmax=1`.

This should also raise a nice error message like we do when the user passes `interpolation` when the user passes `norm` (with this change they will get a "passed the same argument twice" error out of Python which while technically correct, we can give the user a more informative error).  That should also get a test (search `with pytest.raises` for examples of how to do that).

**Comment by bakesbasha** [comment]
I added test_spy.py which compares output of imshow vs spy using the original example on the issue report

**Comment by tacaswell** [comment]
Can you put the new test in `test_image.py` rather than making a new file?  While there is a distinct public API for this, it is a wrapper for `imshow` so it makes sense to me to keep the tests together.

**Comment by bakesbasha** [comment]
Sure, I wasn't sure where to add the test, I'll do that as soon as I can!

**Comment by bakesbasha** [comment]
> Can you put the new test in `test_image.py` rather than making a new file? While there is a distinct public API for this, it is a wrapper for `imshow` so it makes sense to me to keep the tests together.

Fixed

File: lib/matplotlib/tests/test_image.py [hunk 1]
```diff
@@ -1218,3 +1219,69 @@ def test_huge_range_log(fig_test, fig_ref):
...
+
+    # plotting with spy
+    ax_test[0].set_title("ones")
+    ax_test[0].spy([[1, 1], [1, 1], ])
+    ax_test[1].set_title("zeros")
+    ax_test[1].spy([[0, 0], [0, 0], ])
+    ax_test[2].set_title("mixed")
+    ax_test[2].spy([[0, 1], [1, 0], ])
+
+    # plotting with imshow
```
Comment (jklymak) [inline lib/matplotlib/tests/test_image.py:1238]: You have the same code repeated three times here.  Suggest you rewrite as a loop? i.e. 

```
for z, title in zip([[1, 1], [1, 1]], [[0, 0], [0, 0]], [[0, 1], [1, 0]], ['ones', 'zeros', 'both']):
```

File: lib/matplotlib/tests/test_image.py [hunk 2]
```diff
@@ -1218,3 +1219,69 @@ def test_huge_range_log(fig_test, fig_ref):
...
+    ax_ref[2].set_ylim(1.5, -0.5)
+    ax_ref[2].xaxis.tick_top()
+    ax_ref[2].title.set_y(1.05)
+    ax_ref[2].xaxis.set_ticks_position('both')
+    ax_ref[2].xaxis.set_major_locator(
+        mticker.MaxNLocator(nbins=9, steps=[1, 2, 5, 10], integer=True)
+    )
+    ax_ref[2].yaxis.set_major_locator(
+        mticker.MaxNLocator(nbins=9, steps=[1, 2, 5, 10], integer=True)
+    )
```
Comment (jklymak) [inline lib/matplotlib/tests/test_image.py:1285]: This will need to pass flake8 before it can be merged.

Comment (bakesbasha) [reply lib/matplotlib/tests/test_image.py]: Fixed, thanks for letting me know

**Review by jklymak — APPROVED** [review]
Looks good to me.  Thanks!

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -7662,8 +7662,9 @@ def spy(self, Z, precision=0, marker=None, markersize=None,
             if 'interpolation' in kwargs:
                 raise TypeError(
                     "spy() got an unexpected keyword argument 'interpolation'")
-            ret = self.imshow(mask, interpolation='nearest', aspect=aspect,
-                              origin=origin, **kwargs)
+            ret = self.imshow(mask, norm=mcolors.NoNorm(),
+                                interpolation='nearest', aspect=aspect,
+                                origin=origin, **kwargs)
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:7667]:
```suggestion
            ret = self.imshow(mask, norm=mcolors.NoNorm(),
                              interpolation='nearest', aspect=aspect,
                              origin=origin, **kwargs)
```

File: lib/matplotlib/tests/test_image.py [hunk 4]
```diff
@@ -1218,3 +1219,35 @@ def test_huge_range_log(fig_test, fig_ref):
...
+    # setting up reference and test
+    ax_test = fig_test.subplots(1, 3)
+    ax_ref = fig_ref.subplots(1, 3)
+
+    i = 0
+
+    plot_data = [[1, 1], [1, 1], ], [[0, 0], [0, 0], ], [[0, 1], [1, 0], ]
+
+    for z, title in zip((plot_data), ["ones", "zeros", "mixed"]):
+        print(i, z, title)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_image.py:1235]:
```suggestion
    plot_data = (
        [[1, 1], [1, 1]],
        [[0, 0], [0, 0]],
        [[0, 1], [1, 0]],
    )
    plot_titles = ["ones", "zeros", "mixed"]

    for i, (z, title) in enumerate(zip(plot_data, plot_titles)):
```
and delete the `i += 1` at the end.

**Comment by bakesbasha** [comment]
Added some suggested fixes and some styling fixes. Let me know if there is anything else that needs amending!

**Comment by tacaswell** [comment]
Can you please add the validation that the user did not pass us `norm` via kwarg (like the `interpolation` check just above?)?

**Comment by tacaswell** [comment]
A better alternative may be to only provide the default norm if the user did not provide one (to not break the suggested workaround in the original issue!).

**Comment by bakesbasha** [comment]
I removed `norm=mcolors.NoNorm()` as an argument from `ret`, and instead added it to `kwargs` in the case the user didn't pass a `norm`

**Review by tacaswell — APPROVED** [review]
I am confused by why codecove thinks we are missing the line in `_axes.py`

**Comment by QuLogic** [comment]
I think they have an off-by-one error; it's the other side of the `else` that is untested.

**Comment by jklymak** [comment]
... is the codecov problem this PR?

**Comment by tacaswell** [comment]
Thanks @bakesbasha .  Congratulations on your first PR to Matplotlib :tada:   Hopefully we will hear from you again!

**Comment by tacaswell** [comment]
@jklymak No, I think it is a combination of travis dropping off and codecov getting confused about how to track the changes of the diff (the new test code is covered so I am confident that the line in the library is also covered).

**Pull request merged**